subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/transaction/CompileTransaction.java

```diff
@@ -233,15 +231,25 @@ private void moveCompileOutputToOriginalFolders(List<StagedOutput> stagedOutputs
     }
 
     private void rollbackStash(List<StashedFile> stashedFiles) {
-        stashedFiles.forEach(StashedFile::unstash);
+        if (isIncrementalCompilationAfterFailure()) {
+            stashedFiles.forEach(StashedFile::unstash);
+        }
     }
 
     private void setupSpecOutputs(List<StagedOutput> stagedOutputs) {
-        stagedOutputs.forEach(StagedOutput::setupSpecOutput);
+        if (isIncrementalCompilationAfterFailure()) {
+            stagedOutputs.forEach(StagedOutput::setupSpecOutput);
+        }
     }
 
     private void restoreSpecOutputs(List<StagedOutput> stagedOutputs) {
-        stagedOutputs.forEach(StagedOutput::restoreSpecOutput);
+        if (isIncrementalCompilationAfterFailure()) {
+            stagedOutputs.forEach(StagedOutput::restoreSpecOutput);
+        }
+    }
```

Comment:
I see what you mean. Yes, we could split that, that would makes sense. I would do that next time we visit that class maybe. Since we might need to do some other changes, like splitting tests. And I think it's still somehow manageable, since this is used just in these three methods.

What do you think?